Added new LLMs & Emb Models

Karthi-DStech opened the pull request:
## Pull Request Contents

---
♻️ Refactor
✨ Feature
📝 Documentation
---

## Description

This Pull Request enhances the functionality of the codebase by adding support for multiple **LLMs (Large Language Models)** and **Embedding Models**, and provides comprehensive documentation for the changes. The changes include:

1. **LLM Integration**:
    - Added support for multiple LLMs, including `Llama-2-7B`, `Llama-3-8B`, `Mistral-7B`, and others, allowing users to select and work with different models dynamically.
    - Refactored the `get_model` function to support these LLMs and provide error handling for unsupported models.
2. **Embedding Models**:
    - Integrated several new embedding models such as `DistilBERT`, `GTR-T5`, `E5`, and `Contriever` into the system for vector-based document search and retrieval.
3. **Documentation**:
    - Comprehensive documentation has been written for both LLMs and embedding models, explaining their use cases, benefits, and technical specifications.
    - Improved docstrings and inline comments throughout the codebase to enhance clarity and maintainability.

## Added/updated tests?

Test Cases should be Done. Try to run the code and see if we can implement the code without any errors. 

I didn’t make any changes to the existing code and it should work without any trouble.

kuraisle commented:
What are the issue numbers here? We don't have 5678 issues yet!

kuraisle commented on Carrot-Assistant/components/pipeline.py:
Are these just formatting changes?

Karthi-DStech replied:
Yes. It formats the code and looks for common errors in our code. 

(eg: if we are not using an import yet we imported, it will show an error and I've removed it).

kuraisle commented on Carrot-Assistant/components/embeddings.py:
This information is useful, but I don't think it belongs in the code itself. Could you remove this information, then make a separate documentation PR, creating a file in llettuce-docs for the embeddings model. Maybe a nice table like

| Model Name | Version | Dimensions | Description | Documentation |
|---|---|---|---|---|
| Bidirectional Gated Encoder | Small | 384 | etc. | etc. |

I find the code harder to read with this info in. I'm also worried having this in might interfere with the docs build, as the API reference is built from the docstrings

Karthi-DStech replied:
Sure, will implement the same as you requested.

kuraisle reviewed (changes requested):
Thanks for adding the info for the different models we can experiment with.

My concern is having so much info in the code. I find it makes it harder to read. Do you think you could move it to a page in the documentation? Having it formatted for the web there will be easier to figure out and compare models if it's in a table, and it'll be easier to get to the code to figure out what's going on. Does that make sense?

Karthi-DStech commented:
Yep, I get it. For the time being, I'll remove the extra comments, and we can create separate documentation of models and provide detailed information.

Karthi-DStech commented:
> What are the issue numbers here? We don't have 5678 issues yet!

That's a typo and I've fixed it.

kuraisle commented:
Why not something like this?

https://qdrant.github.io/fastembed/examples/Supported_Models/

I just don't think that someone using lettuce needs to go through our code to find which models they can use. We have an API that means they can make HTTP requests with the model as an option. I think it's good documentation to have the information easy to find, not buried in the code

Karthi-DStech commented:
> Why not something like this?
> 
> https://qdrant.github.io/fastembed/examples/Supported_Models/
> 
> I just don't think that someone using lettuce needs to go through our code to find which models they can use. We have an API that means they can make HTTP requests with the model as an option. I think it's good documentation to have the information easy to find, not buried in the code

Alright James. I've removed the comments and pushed into a separate file and referenced it. Please review it and let me know for further changes.

kuraisle commented on docs/_sources/autoapi/OMOP_match/embedding_models.rst.txt:
the docs folder is where the docs are built. It would be great if the .rst could go in /llettuce-docs

kuraisle commented on docs/_sources/autoapi/OMOP_match/embedding_models.rst.txt:
```diff
@@ -0,0 +1,307 @@
+        is based on T5 and optimized for generalization 
+        across diverse tasks.
+
+    - Benifits:
```
also Ben*e*fits

Karthi-DStech commented:
Hi James, 

I have updated the models and tested everything. I've also removed some models because they throw some errors and some don't have GGUF format. All the models are ready to predict and work well without any errors. 

Also, I've moved the .rst file to lettuce docs. Kindly do the needful.

kuraisle commented on Carrot-Assistant/components/models.py:
```diff
@@ -49,39 +139,565 @@ def get_model(
+        # ------ Llama Model (7B) ------ >
+        # --------------------------
+
+        elif model_name == "llama2-7b":
```
It would be less verbose to use hf_hub_download(**local_models[model_name]) than this elif chain

Karthi-DStech replied:
Sure

kuraisle commented on llettuce-docs/embedding_models.rst:
```diff
@@ -0,0 +1,307 @@
+        embedding, semantic similarity, and information retrieval. 
+        
+    - Benefits:
+        BGE-small more efficient in terms of speed and memory,
```
More efficient than?

kuraisle commented on Carrot-Assistant/components/models.py:
```diff
@@ -49,39 +139,565 @@ def get_model(
 
-        try:
-            llm = LlamaCppGenerator(
-                model=hf_hub_download(**local_models[model_name]),
```
Just put it back to this and it should still work

Karthi-DStech replied:
Sure. I'll work on it.

kuraisle commented on llettuce-docs/embedding_models.rst:
```diff
@@ -0,0 +1,307 @@
+        
+        It leverages the T5 model for deep semantic understanding and 
+        ensures fast retrieval using Approximate Nearest Neighbor 
+        (ANN) search, making it both powerful and efficient 
```
I don't think we're implementing ANN, at least yet

kuraisle commented on llettuce-docs/embedding_models.rst:
```diff
@@ -0,0 +1,307 @@
+
+
+
+4. Generalizable T5 Retrieval (Large)
```
If some of the models are different sized versions of one another, maybe run them as subheadings?

kuraisle reviewed (approved):
Thanks for putting the new models in
